cpp/src/lists/lists_column_factories.cu

```diff
@@ -119,11 +119,18 @@ std::unique_ptr<column> make_lists_column(size_type num_rows,
                                          null_count,
                                          std::move(children));
 
+  // Check not having nulls using `null_count==0` along with nullable because we may have
+  // `null_count==UNKNOWN_NULL_COUNT`.
+  if (null_count == 0 || !output->nullable() || child_type_id == type_id::EMPTY) { return output; }
+
+  auto const output_cv = output->view();
+
   // We need to enforce all null lists to be empty.
-  // Checking `null_count==0` is not enough as it can be `UNKNOWN_NULL_COUNT` for nullable column.
-  return null_count == 0 || !output->nullable() || child_type_id == type_id::EMPTY
-           ? std::move(output)
-           : detail::purge_nonempty_nulls(output->view(), stream, mr);
+  // `has_nonempty_nulls` is less expensive than `purge_nonempty_nulls` and can save some
+  // performance if we don't have any non-empty nulls.
+  return detail::has_nonempty_nulls(output_cv, stream)
+           ? detail::purge_nonempty_nulls(output_cv, stream, mr)
+           : std::move(output);
```

Comment:
optional suggestion: `has_nonempty_nulls` could also be an early return so we wouldn't need `std::move` or the ternary